module/Company/src/Company/Service/Company.php

```diff
@@ -183,87 +207,193 @@ public function getCategoryList($visible)
-            $id = -1;
-            foreach ($newCategories as $lang => $category) {
-                $arr[$lang] = $this->insertCategory($lang, $id, $category);
-                if ($id == -1) {
-                    $id = current($arr)->getId();
-                }
+        $categoryForm->bind($categories);
+        $categoryForm->setData($data);
+        
+        if (!$categoryForm->isValid()) {
+            return false;
+        }
+
+        $id = -1;
+        foreach ($categories as $category) {
+            $id = $this->setLanguageNeutralCategoryId($id, $category, $languageNeutralId);
+            $this->getCategoryMapper()->persist($category);
+            $this->saveCategory();
+        }
+
+        return (($languageNeutralId == "") ? $id : $languageNeutralId);
+    }
+
+    /**
+     * Sets the languageNeutralId for this JobCategory.
+     *
+     * @param int $id The id of the JobCategory
+     * @param JobCategory $category The JobCategory
+     * @param int|string $languageNeutralId The languageNeutralId of the JobCategory
+     *
+     * @return int
+     */
+    private function setLanguageNeutralCategoryId($id, $category, $languageNeutralId)
+    {
+        if ($languageNeutralId == "") {
+            $category->setLanguageNeutralId($id);
+            $this->getCategoryMapper()->persist($category);
+            $this->saveCategory();
+
+            if ($id == -1) {
+                $id = $category->getId();
             }
-            return $arr;
+
+            $category->setLanguageNeutralId($id);
+            return $id;
         }
-        return null;
+
+        $category->setLanguageNeutralId($languageNeutralId);
+        return $id;
+    }
+
+    /**
+     * Creates a new JobLabel.
+     *
+     * @param array $data Label data from the EditLabel form
+     * @return bool|int Returns false on failure, and the languageNeutralId on success
+     */
+    public function createLabel($data)
+    {
+        $labelDict = [];
+        foreach ($this->getLanguages() as $lang) {
+            $label = new LabelModel();
+            $label->setLanguage($lang);
+            $labelDict[$lang] = $label;
+        }
+
+        return $this->saveLabelData("", $labelDict, $data);
+    }
+
+    /**
+     * Checks if the data is valid, and if it is, saves the JobLabel
+     *
+     * @param int|string $languageNeutralId Identifier of the JobLabel to save
+     * @param array $labels The JobLabels to save
+     * @param array $data The data to validate, and apply to the label
+     *
+     * @return bool|int
+     */
+    public function saveLabelData($languageNeutralId, $labels, $data)
+    {
+        $labelForm = $this->getLabelForm();
+        $labelForm->bind($labels);
+        $labelForm->setData($data);
+        
+        if (!$labelForm->isValid()) {
+            return false;
+        }
+
+        $id = -1;
+        foreach ($labels as $label) {
+            $id = $this->setLanguageNeutralLabelId($id, $label, $languageNeutralId);
```

Comment:
Why is the ID set to -1 and then overwritten multiple times in the foreach loop?

Reply:
This code is a modified version of the `saveJobData` function, which is from (at least) 4 years ago. We might have to refactor the label and category functions as they are virtually the same and combine them. But until we switch to PHP 7 this will be messy.

If required we can change the behaviour, but we also have to change it for the aforementioned job function.

Reply:
To answer your question, on label/category creation, there is no identifier for the item. As the item is created, the `$languageNeutralId` is empty (perhaps `null` is better). The first identifier is created and will become the `$languageNeutralId`, after which the `$id` is used to update the `$languageNeutralId` for the second item (i.e. language).